pkg/kubelet/kubelet_test.go

```diff
@@ -3657,3 +3657,145 @@ func TestSyncPodsWithRestartPolicy(t *testing.T) {
+			},
+		},
+		"5678": {
+			ID:     "5678",
+			Name:   "failed",
+			Config: &docker.Config{},
+			State: docker.State{
+				ExitCode:   42,
+				StartedAt:  time.Now(),
+				FinishedAt: time.Now(),
+			},
+		},
+	}
+
+	tests := []struct {
+		policy           api.RestartPolicy
+		calls            []string
+		created          []string
+		stopped          []string
+		lastTerminations []string
+	}{
+		{
+			api.RestartPolicyAlways,
+			[]string{"list", "list", "list", "inspect_container", "inspect_container", "inspect_container", "create", "start", "create", "start", "list", "inspect_container", "inspect_container", "inspect_container", "inspect_container", "inspect_container"},
+			[]string{"succeeded", "failed"},
+			[]string{},
+			[]string{"docker://1234", "docker://5678"},
+		},
+		{
+			api.RestartPolicyOnFailure,
+			[]string{"list", "list", "list", "inspect_container", "inspect_container", "inspect_container", "create", "start", "list", "inspect_container", "inspect_container", "inspect_container", "inspect_container"},
+			[]string{"failed"},
+			[]string{},
+			[]string{"docker://5678"},
+		},
+		{
+			api.RestartPolicyNever,
+			[]string{"list", "list", "list", "inspect_container", "inspect_container", "inspect_container", "stop", "list", "inspect_container", "inspect_container"},
+			[]string{},
+			[]string{"9876"},
+			[]string{},
+		},
+	}
+
+	for i, tt := range tests {
+		fakeDocker.ContainerList = runningAPIContainers
+		fakeDocker.ExitedContainerList = exitedAPIContainers
+		fakeDocker.ContainerMap = containerMap
+		fakeDocker.ClearCalls()
+		pods := []api.Pod{
+			{
+				ObjectMeta: api.ObjectMeta{
+					UID:       "12345678",
+					Name:      "foo",
+					Namespace: "new",
+				},
+				Spec: api.PodSpec{
+					Containers:    containers,
+					RestartPolicy: tt.policy,
+				},
+			},
+		}
+		kubelet.podManager.SetPods(pods)
+		waitGroup.Add(1)
+		err := kubelet.SyncPods(pods, emptyPodUIDs, map[string]api.Pod{}, time.Now())
+		if err != nil {
+			t.Errorf("%d: unexpected error: %v", i, err)
+		}
+		waitGroup.Wait()
+
+		// Check if we can retrieve the pod status from GetPodStatus().
+		podName := kubecontainer.GetPodFullName(&pods[0])
+		status, err := kubelet.GetPodStatus(podName)
+		if err != nil {
+			t.Fatalf("unable to retrieve pod status for pod %q: %#v.", podName, err)
+		} else {
+			terminatedContainers := []string{}
+			for _, cs := range status.ContainerStatuses {
+				if cs.LastTerminationState.Termination != nil {
+					terminatedContainers = append(terminatedContainers, cs.LastTerminationState.Termination.ContainerID)
+				}
+			}
+			sort.StringSlice(terminatedContainers).Sort()
+			sort.StringSlice(tt.lastTerminations).Sort()
+			if !reflect.DeepEqual(terminatedContainers, tt.lastTerminations) {
+				t.Errorf("Expected(sorted): %#v, Actual(sorted): %#v", tt.lastTerminations, terminatedContainers)
+			}
+		}
+
+		// 'stop' is because the pod infra container is killed when no container is running.
```

Comment:
WDYT of not checking this and only checking pod status? Should keep the test smaller and more focused. Do we event need to verifyCalls?